### PR TITLE
[Validator] Property paths starting with 0 are broken.

### PR DESCRIPTION
**Pull request by fago**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14394 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/Validator/Util/PropertyPath.php [hunk 1]
```diff
@@ -42,7 +42,7 @@ public static function append($basePath, $subPath)
                 return $basePath.$subPath;
             }
 
-            return $basePath ? $basePath.'.'.$subPath : $subPath;
+            return '' !== (string) $basePath ? $basePath.'.'.$subPath : $subPath;
```
Comment (staabm) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: I think you wont need the explicit  string-cast when strictly comparing against `''`

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php [hunk 2]
```diff
@@ -590,9 +590,7 @@ private function validateClassNode($object, $cacheKey, ClassMetadataInterface $m
                     $object,
                     $cacheKey.':'.$propertyName,
                     $propertyMetadata,
-                    $propertyPath
-                        ? $propertyPath.'.'.$propertyName
-                        : $propertyName,
+                    PropertyPath::append($propertyPath, $propertyName),
```
Comment (staabm) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: This should deservere a separate unit test

File: src/Symfony/Component/Validator/Util/PropertyPath.php (hunk 1 above)
Comment (fago) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: $basePath could be FALSE, or NULL also - in which case we should not add the point delimiter also.

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php (hunk 2 above)
Comment (fago) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: Makes sense, it doesn't look like there is a unit test for RecursiveContextualValidator already though?

Comment (dawehner) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: \Symfony\Component\Validator\Tests\Validator\RecursiveValidator2Dot5ApiTest is the one or rather the parent abstract class

File: src/Symfony/Component/Validator/Util/PropertyPath.php (hunk 1 above)
Comment (Tobion) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: no, base path is not allowed to be false or null. see phpdoc

**Comment by stof** [comment]
:-1: on this change, as its goal is to allow using an invalid property path. The drupal code should be fixed instead.

**Comment by dawehner** [comment]
Fair, for the usecase of symfony and custom PHP classes, this is not what you wanna do.
On the other hand, properties are an abstract concept, not necessarily bound to concrete classes.

I guess the way to fix it would be to use traversables for that part of our property path.

On the other hand I wish we would have communicated when the symfony validator was rewritten

File: src/Symfony/Component/Validator/Util/PropertyPath.php (hunk 1 above)
Comment (fago) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: Neither is subpath - still it is check already that way in the same method.

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php (hunk 2 above)
Comment (fago) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: Right, I got confused with "UnitTest" - as this is not really a unit test. But yeah ok, we should be able to extend this one with coverage for that case.

**Comment by fago** [comment]
I don't see why 0.value would be an invalid property path? Where does it say numbers are not allowed? 
On the contrary, when numerically indexed arrays are validated you'll always end up with numbers in the property path?

File: src/Symfony/Component/Validator/Util/PropertyPath.php (hunk 1 above)
Comment (Tobion) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: if base path is not a string, then the method would also not return a string. so it must be a string.

Comment (Tobion) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: or you have to cast the return value as well.  but that is not what we want.

**Comment by webmozart** [comment]
@stof Why? To me this seems like a simple bug fix.

**Comment by Tobion** [comment]
Agree, it's a simple bug fix. That the goal is a wrong use-case doesn't matter.

File: src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php (hunk 2 above)
Comment (webmozart) [inline src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php]: I don't think we need another unit test here. The two cases of the current logic are (or at least should be) already tested. Exchanging the current explicit logic by a method call should not require additional tests, as we don't change the functionality. The internal behavior of `append()` is tested elsewhere.

File: src/Symfony/Component/Validator/Util/PropertyPath.php (hunk 1 above)
Comment (webmozart) [inline src/Symfony/Component/Validator/Util/PropertyPath.php]: I think the change is fine given the current implementation. If we want to make the method stricter, i.e. check the input more rigorously, that's the subject of another PR.

**Comment by webmozart** [comment]
ping @stof

**Comment by webmozart** [comment]
No response from @stof. I'm going to merge this.

**Comment by webmozart** [comment]
Thank you @fago.

**Comment by jibran** [comment]
Thank you @webmozart

**Comment by Tobion** [comment]
As a bug fix this should have been merged into 2.6

**Comment by fabpot** [comment]
I've cherry-picked the commit on 2.6.

**Pull request closed without merging**